src/lib/keychain/insecure.js

```diff
@@ -35,33 +47,56 @@ export default class Secure implements Keychain {
 
 		// Ensure permissions are what we expect
 		if ( !! ( perms & ~rw ) ) {
-			throw 'Invalid permissions on access token file: ' + tmpfile;
+			throw `Invalid permissions (${ perms.toString( 8 ) }, expecting ${ rw.toString( 8 ) }) for keychain file (${ tmpfile })`;
 		}
 
 		this.file = tmpfile;
 	}
 
 	getPassword( service: string ): Promise<string> {
+		if ( this.passwords && this.passwords[ service ] ) {
+			return Promise.resolve( this.passwords[ service ] );
+		}
+
 		return new Promise( resolve => {
-			fs.readFile( this.file, 'utf8', ( err, password ) => {
-				if ( err || ! password ) {
+			fs.readFile( this.file, 'utf8', ( err, passwords ) => {
+				if ( err || ! passwords ) {
 					return resolve( null );
 				}
 
-				return resolve( password );
+				this.passwords = JSON.parse( passwords );
+
+				return resolve( passwords[ service ] );
 			} );
 		} );
 	}
 
 	setPassword( service: string, password: string ): Promise<boolean> {
+		this.passwords[ service ] = password;
+
 		return new Promise( resolve => {
-			fs.writeFile( this.file, password, err => resolve( ! err ) );
+			const json = JSON.stringify( this.passwords );
+			fs.writeFile( this.file, json, err => resolve( ! err ) );
```

Comment:
Would this (and `delete` below) fail silently? Should we surface the error somehow?